File: rust/src/models/subscription_component.rs
// this file is @generated
use serde::{Deserialize, Serialize};

use super::{
    price_component_id::PriceComponentId, product_id::ProductId, subscription_fee::SubscriptionFee,
    subscription_fee_billing_period_enum::SubscriptionFeeBillingPeriodEnum,
};

#[derive(Clone, Debug, Default, PartialEq, Deserialize, Serialize)]
pub struct SubscriptionComponent {
    pub fee: SubscriptionFee,

    pub name: String,

    pub period: SubscriptionFeeBillingPeriodEnum,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub price_component_id: Option<PriceComponentId>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub product_id: Option<ProductId>,
}

impl SubscriptionComponent {
    pub fn new(
        fee: SubscriptionFee,
        name: String,
        period: SubscriptionFeeBillingPeriodEnum,
    ) -> Self {
        Self {
            fee,
            name,
            period,
            price_component_id: None,
            product_id: None,
        }
    }
}

File: typescript/src/models/subscriptionComponent.ts
// this file is @generated
import { type PriceComponentId, PriceComponentIdSerializer } from "./priceComponentId";
import { type ProductId, ProductIdSerializer } from "./productId";
import { type SubscriptionFee, SubscriptionFeeSerializer } from "./subscriptionFee";
import {
  type SubscriptionFeeBillingPeriodEnum,
  SubscriptionFeeBillingPeriodEnumSerializer,
} from "./subscriptionFeeBillingPeriodEnum";

export interface SubscriptionComponent {
  fee: SubscriptionFee;

  name: string;

  period: SubscriptionFeeBillingPeriodEnum;

  priceComponentId?: PriceComponentId | null;

  productId?: ProductId | null;
}

export const SubscriptionComponentSerializer = {
  _fromJsonObject(object: any): SubscriptionComponent {
    return {
      fee: SubscriptionFeeSerializer._fromJsonObject(object["fee"]),
      name: object["name"],
      period: SubscriptionFeeBillingPeriodEnumSerializer._fromJsonObject(
        object["period"]
      ),
      priceComponentId:
        object["price_component_id"] != null
          ? PriceComponentIdSerializer._fromJsonObject(object["price_component_id"])
          : undefined,
      productId:
        object["product_id"] != null
          ? ProductIdSerializer._fromJsonObject(object["product_id"])
          : undefined,
    };
  },

  _toJsonObject(self: SubscriptionComponent): any {
    return {
      fee: SubscriptionFeeSerializer._toJsonObject(self.fee),
      name: self.name,
      period: SubscriptionFeeBillingPeriodEnumSerializer._toJsonObject(self.period),
      price_component_id:
        self.priceComponentId != null
          ? PriceComponentIdSerializer._toJsonObject(self.priceComponentId)
          : undefined,
      product_id:
        self.productId != null
          ? ProductIdSerializer._toJsonObject(self.productId)
          : undefined,
    };
  },
};
